include/seastar/core/semaphore.hh

```diff
@@ -364,6 +364,21 @@ public:
             return false;
         }
     }
+
+    /// Returns true iff the specified number of units can be reduced from the counter without waiting.
+    ///
+    /// If sufficient units are available in the counter, and if no
+    /// other fiber is waiting, then the function returns true.  Otherwise,
+    /// nothing happens.  This is useful for "opportunistic" waits where
+    /// useful work can happen if the counter happens to be ready, but
+    /// when it is not worthwhile to wait.
+    ///
+    /// \param nr number of units to reduce the counter by (default 1).
+    /// \return `true` if the counter had sufficient units and no other fiber is waiting.
+    bool can_acquire_immediately(size_t nr = 1) noexcept {
```

Comment:
Likely deserves `const` qualifier

Reply:
posix threads has _trylock() methods, which return an error if the lock can not be acquired immediately. With a boolean check like this it takes two calls to do the same. Perhaps _trylock() api should be preferred.

Reply:
Posix doesn't have a method that would return the state of the lock deliberately. First this information is immediately obsolete, since the lock may be released right at the moment. It's true in Scylla the information stays fresh until a yield point - still the code that relies on it becomes fragile.

The second reason is that the state of the lock is not the same as the state of the locked object. I remember a relevant discussion at comp.programming.threads where  David M. Butenhof, one of the authors of the original HP implementation and then member of the standards committee, was advocating that locks are not part of the state, they are used to protect that state. In other words, this becomes fragile because holding (or more importantly - not holding) the lock one day may stop reflecting the invariant it was used to check for.

Reply:
This akin to the atomic_read part of test_and_set.
I see no reason why not allow it as a primitive, to test the negative case of when the lock is already taken, just without the side effects of try_lock (when it succeeds)